news-explorer-frontend/src/components/PopupWithForm/PopupWithForm.js

```diff
@@ -0,0 +1,310 @@
+import React, { useEffect, useState } from 'react';
+import { Link, useNavigate } from 'react-router-dom';
+import closebtn from '../../images/close.svg';
+import { authorize, register } from '../../utils/auth';
+import './PopupWithForm.css';
+
+function PopupWithForm(props) {
+  const isSignInPopupOpen = props.isSignInPopupOpen;
+  const isRegisterPopupOpen = props.isRegisterPopupOpen;
+  const formIsOpen = (isSignInPopupOpen || isRegisterPopupOpen) ? 'popup_open' : '';
+  const users = props.users;
+  const [SubmitButtonClass, setSubmitButtonClass] = useState('popup__submit-btn_disabled');
+  const [nameInputfield, setNameInputField] = useState({
+    field: '',
+    value: '',
+    error: '',
+    isValid: false,
+  });
+  const [emailInputfield, setEmailInputField] = useState({
+    field: '',
+    value: '',
+    error: '',
+    isValid: false,
+  });
+  const [passwordInputfield, setPasswordInputField] = useState({
+    field: '',
+    value: '',
+    error: '',
+    isValid: false,
+  });
+  const [formErrorMessage, setFormErrorMessage] = useState({
+    message: '',
+    formName: '',
+  });
+
+  const signinFormValid = emailInputfield.isValid && passwordInputfield.isValid;
+  const signupFormValid = emailInputfield.isValid && passwordInputfield.isValid && nameInputfield.isValid;
+
+  const [formIsValid, setFormIsValid] = useState(false);
+  const navigate = useNavigate();
+
+  const toggleBtnState = () => {
+    if (formIsValid) {
+      setSubmitButtonClass('');
+    }
+    if (!formIsValid) {
+      setSubmitButtonClass('popup__submit-btn_disabled');
+    }
+  };
+
+  useEffect(() => {
+    setFormIsValid(isSignInPopupOpen ? signinFormValid : signupFormValid);
+  }, [signinFormValid, signupFormValid]);
+
+  useEffect(() => {
+    toggleBtnState();
+  }, [formIsValid]);
+
+  const close = () => {
+    props.onClosePopup();
+    navigate('/');
+  };
+
+  const handleClickOnOverlayClose = () => {
+    close();
+  };
+
+  const handleSigninSubmit = (email, password) => {
```

Comment:
`PopupWithForm` should be a universal component. Submit handlers should be passed to the `props`